### PR TITLE
C++: IR construction for lambda expressions

### PR DESCRIPTION
**Pull request by dave-bartolomeo**
The IR construction code wasn't handling lambda expressions, so I added `TranslatedLambdaExpression`. It's pretty straightforward: it creates a temporary variable, initializes it with an `Uninitialized` instruction, then initializes the individual captured fields with the initializer list supplied in the AST.

When testing the case of a lambda with no captures, I noticed that we weren't handling initialization of empty structs with an initializer list correctly, so I fixed that along the way.

I was getting confused by the bad indentation for wrapped lines in
TranslatedInitialization.qll, so I fixed that up in a separate commit.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll [hunk 1]
```diff
@@ -268,6 +268,9 @@ newtype TTranslatedElement =
       ) or
       exists(ThrowExpr throw |
         throw.getExpr().getFullyConverted() = expr
+      ) or
+      exists(LambdaExpression lambda |
+        lambda.getChild(0).getFullyConverted() = expr
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll]: If `getChild(0)` is a useful value to query for a `LambdaExpression`, there should be a getter for it.

File: cpp/ql/test/library-tests/ir/ir/ir.cpp [hunk 2]
```diff
@@ -1006,4 +1006,25 @@ void OperatorDeleteArray() {
...
+  auto lambda_empty = [](float f) { return 'A'; };
+  lambda_empty(0);
+  auto lambda_ref = [&](float f) { return s.c_str()[x]; };
+  lambda_ref(1);
+  auto lambda_val = [=](float f) { return s.c_str()[x]; };
+  lambda_val(2);
+  auto lambda_ref_explicit = [&s](float f) { return s.c_str()[0]; };
+  lambda_ref_explicit(3);
+  auto lambda_val_explicit = [s](float f) { return s.c_str()[0]; };
+  lambda_val_explicit(4);
```
Comment (jbj) [inline cpp/ql/test/library-tests/ir/ir/ir.cpp]: Can you add a test with multiple captured variables, ideally with different capture modes. Is the translation intended to support [the full range of C++17 capture modes](https://en.cppreference.com/w/cpp/language/lambda#Lambda_capture)?

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll (hunk 1 above)
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll]: Added `LambdaExpression.getInitializer()`

File: cpp/ql/test/library-tests/ir/ir/ir.cpp (hunk 2 above)
Comment (dave-bartolomeo) [reply cpp/ql/test/library-tests/ir/ir/ir.cpp]: Done. Captures with initializers, which I did not know existed until now, work with no changes. Explicit captures of `this` produce some incorrect-looking ASTs, so I'll have to dig into that separately. I've added an `#ifd'd-out test case for those, and will investigate further after this PR is merged.

**Comment by dave-bartolomeo** [comment]
@rdmarsh2 Now that I've fixed all of @jbj's feedback, can you do a review and merge?

**Pull request merged**